skrifa/src/color/traversal.rs

```diff
@@ -52,10 +52,76 @@ fn make_sorted_resolved_stops(stops: &ColorStops, instance: &ColrInstance) -> Ve
     collected
 }
 
-pub(crate) fn traverse_with_callbacks(
+struct CollectFillGlyphPainter<'a> {
+    brush_transform: Transform,
+    glyph_id: GlyphId,
+    optimizable: bool,
+    parent_painter: &'a mut dyn ColorPainter<'a>,
+    result: Result<FillGlyph, PaintError>,
+}
+
+impl<'a> CollectFillGlyphPainter<'a> {
+    fn new(parent_painter: &'a mut impl ColorPainter<'a>, glyph_id: GlyphId) -> Self {
+        Self {
+            brush_transform: Transform::default(),
+            glyph_id,
+            parent_painter,
+            optimizable: true,
+            result: Ok(FillGlyph::Ok),
+        }
+    }
+
+    fn paint_successful(&self) -> bool {
+        self.result.is_ok()
+    }
+}
+
+impl<'a> ColorPainter<'a> for CollectFillGlyphPainter<'a> {
+    fn push_transform(&mut self, transform: Transform) {
+        self.brush_transform *= transform;
+    }
+
+    fn pop_transform(&mut self) {
+        // Since we only support fill and and transform operations, we need to
+        // ignore a popped transform, as this would be called after traversing
+        // the graph backup after a fill was performed, but we want to preserve
+        // the transform in order to be able to return it.
+    }
+
+    fn fill<'b>(&mut self, brush: Brush<'b>) {
+        if !self.optimizable {
+            return;
+        }
+        self.result =
+            self.parent_painter
+                .fill_glyph(self.glyph_id, self.brush_transform.clone(), brush);
+    }
+
+    fn push_clip_glyph(&mut self, _: GlyphId) {
+        self.optimizable = false;
+    }
+
+    fn push_clip_box(&mut self, _: BoundingBox<f32>) {
+        self.optimizable = false;
+    }
+
+    fn pop_clip(&mut self) {
+        self.optimizable = false;
+    }
+
+    fn push_layer(&mut self, _: CompositeMode) {
+        self.optimizable = false;
+    }
+
+    fn pop_layer(&mut self) {
+        self.optimizable = false;
+    }
+}
+
+pub(crate) fn traverse_with_callbacks<'a>(
     paint: &ResolvedPaint,
     instance: &ColrInstance,
-    painter: &mut impl ColorPainter,
+    painter: &mut impl ColorPainter<'a>,
```

Comment:
Can y'all help me with the lifetimes and tagging them correctly here? Or is what I am doing even possible. 

If I don't put `&'a mut impl ColorPainter<'a>` here, i.e. no explicit borrow lifetime, than the compiler complains with:

```
434 |                 let mut optimizer = CollectFillGlyphPainter::new(painter, *glyph_id);
    |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ lifetime `'a` required
```

If I add it, I get a lot of errors about not being able to reborrow painter - which previously worked. 

I kind of understand the first error message, but I don't understand how adding an explicit borrow lifetime breaks everything. 

What could be a solution?

Reply:
If you want to take this approach I think what you need to do is have two lifetimes, with an explicit bound. See https://doc.rust-lang.org/1.75.0/reference/trait-bounds.html#lifetime-bounds.

Here that would look like,


```rust
PainterPainter<'inner, 'ourborrow> where 'inner: 'ourborrow {
    painter: &'ourborrow Painter<'inner>,
}
```

Reply:
Thanks so much, that did the trick. I would not have figured that out so soon. Would a prose way to describe this as follows be correct?
> `PainterPainter` is an object that's constructed with an own lifetime of `ourborrow` that borrows a `Painter` struct for the lifetime of `ourborrow`, but requires that `Painter` struct with own lifetime `inner` to outlive `ourborrow` (`PainterPainter`s own lifetime)?

Reply:
Thanks both for your help with that - updated change solves the lifetime issues.